### PR TITLE
Merge Show Profile Picture

### PR DESCRIPTION
**Pull request by agaldanaw2**
GitHub Issue (If applicable): #136 
## PR Type

- Feature

## What is the current behavior?

The app doesn't show the current profile picture of user

## What is the new behavior?

Show current user profile picture

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Tested code with current [supported SDKs](../README.md#supported)
- [ ] Docs have been added/updated which fit [documentation template](https://github.com/nventive/Uno/blob/master/doc/.feature-template.md). (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](doc/articles/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] [Wasm UI Tests](doc/articles/working-with-the-samples-apps.md#running-the-webassembly-ui-tests-snapshots) are not showing unexpected any differences. Validate PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Updated the [Release Notes](https://github.com/nventive/Uno/tree/master/doc/ReleaseNotes)
- [ ] Associated with an issue (GitHub or internal)

### CODE REVIEW

**Review by carldebilly — COMMENTED** [review]
Why did you close https://github.com/unoplatform/uno.todo/pull/232 and create a new PR?

You lost the suggestions I did in the other PR :-(

File: src/ToDo.UI/Converters/BitmapSourceConverter.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+	public object? Convert(object value, Type targetType, object parameter, string language)
+	{
+		if (value is not byte[] data) return null;
+		using var buffer = new MemoryStream(data);
+		var bitmap = new BitmapImage();
+		bitmap.SetSource(buffer.AsRandomAccessStream());
+		return bitmap;
+	}
+	public object ConvertBack(object value, Type targetType, object parameter, string language)
+		=> throw new NotImplementedException("Only one-way conversion is supported.");
```
Comment (carldebilly) [inline src/ToDo.UI/Converters/BitmapSourceConverter.cs:18]:
```suggestion
		=> throw new NotSupportedException("Only one-way conversion is supported.");
```

File: src/ToDo/Presentation/WelcomeViewModel.cs [hunk 2]
```diff
@@ -4,26 +4,30 @@
...
 	private async ValueTask DoGetStarted(CancellationToken ct)
 	{
 		var user = await _authService.AuthenticateAsync(_dispatcher);
-
 		if(user is not null)
 		{
+			var profilePicture = await _userSvc.GetAsync(cancellationToken: ct);
+			_authService.SetUserProfilePicture(profilePicture);
 			await _navigator.NavigateRouteAsync(this, string.Empty, cancellation: ct);
 		}
```
Comment (carldebilly) [inline src/ToDo/Presentation/WelcomeViewModel.cs:32]: Please create a helper method in the AuthService to do that. The setting of the profile picture data should not be done in a ViewModel.

**Comment by jeromelaban** [comment]
FYI, now that the UI tests are run, we can see that this PR introduces a regression in the mocks somehow:

```
05/10/2022 00:17:47/Severe: console-api 0:8 "fail: Uno.Extensions.Reactive.AsyncCommand[0]
      Failed to execute command.
System.InvalidOperationException: Command 'GetStarted' failed when executing command with '-null-'.
 ---> System.Text.Json.JsonException: ExpectedStartOfValueNotFound, \u003C Path: $ | LineNumber: 0 | BytePositionInLine: 0.
 ---> System.Text.Json.JsonReaderException: ExpectedStartOfValueNotFound, \u003C LineNumber: 0 | BytePositionInLine: 0.
   at System.Text.Json.ThrowHelper.ThrowJsonReaderException(Utf8JsonReader& json, ExceptionResource resource, Byte nextByte, ReadOnlySpan`1 bytes)
   at System.Text.Json.Utf8JsonReader.ConsumeValue(Byte marker)
   at System.Text.Json.Utf8JsonReader.ReadFirstToken(Byte first)
   at System.Text.Json.Utf8JsonReader.ReadSingleSegment()
   at System.Text.Json.Utf8JsonReader.Read()
   at System.Text.Json.Serialization.JsonConverter`1[[System.String, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
   Exception_EndOfInnerExceptionStack
   at System.Text.Json.ThrowHelper.ReThrowWithPath(ReadStack& state, JsonReaderException ex)
   at System.Text.Json.Serialization.JsonConverter`1[[System.String, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
   at System.Text.Json.Serialization.JsonConverter`1[[System.String, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].ReadCoreAsObject(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
   at System.Text.Json.JsonSerializer.ReadCore[Object](JsonConverter jsonConverter, Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
   at System.Text.Json.JsonSerializer.ReadCore[Object](JsonReaderState& readerState, Boolean isFinalBlock, ReadOnlySpan`1 buffer, JsonSerializerOptions options, ReadStack& state, JsonConverter converterBase)
   at System.Text.Json.JsonSerializer.ContinueDeserialize[Object](ReadBufferState& bufferState, JsonReaderState& jsonReaderState, ReadStack& readStack, JsonConverter converter, JsonSerializerOptions options)
   at System.Text.Json.JsonSerializer.ReadAll[Object](Stream utf8Json, JsonTypeInfo jsonTypeInfo)
   at System.Text.Json.JsonSerializer.ReadAllUsingOptions[Object](Stream utf8Json, Type returnType, JsonSerializerOptions options)
   at System.Text.Json.JsonSerializer.Deserialize(Stream utf8Json, Type returnType, JsonSerializerOptions options)
   at Uno.Extensions.Serialization.SystemTextJsonSerializer.FromStream(Stream source, Type targetType)
   at Uno.Extensions.Serialization.SerializerExtensions.FromStream[String](ISerializer serializer, Stream stream)
   at Uno.Extensions.Serialization.SystemTextJsonGeneratedSerializer`1[[System.String, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].FromStream(Stream source)
   at Uno.Extensions.Serialization.SystemTextJsonGeneratedSerializer`1[[System.String, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].FromStream(Stream source, Type targetType)
   at Uno.Extensions.Serialization.SerializerExtensions.FromStream[String](ISerializer serializer, Stream stream)
   at Uno.Extensions.Serialization.StorageExtensions.\u003CReadFileAsync>d__0`1[[System.String, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].MoveNext()
   at ToDo.Data.Mock.MockUserProfilePictureEndpoint.\u003CGetAsync>d__4.MoveNext()
   at ToDo.Business.UserProfilePictureService.\u003CGetAsync>d__2.MoveNext()
   at ToDo.Business.Mock.MockAuthenticationService.\u003CSetUserProfilePicture>d__5.MoveNext()
   at ToDo.Presentation.WelcomeViewModel.\u003CDoGetStarted>d__7.MoveNext()
   at Uno.Extensions.Reactive.AsyncCommand.SubCommand.\u003C>c__DisplayClass6_0.\u003C\u003CTryExecute>b__0>d.MoveNext()
   Exception_EndOfInnerExceptionStack"
```

/cc @carldebilly

**Comment by agaldanaw2** [comment]
@jeromelaban got it, but was is it the regresion with the mock? I run the test on my local using main branch and my branch and I got same results: 
image
Can you explain to me or help me to solve the issue? does the issue only happen on local enviroment?

**Comment by Xiaoy312** [comment]
@agaldanaw2 You are running into that NRException locally because the app couldnt find "WelcomePage_Button_Wide".
Try to replace that line with "WelcomePage_Button_Narrow" https://github.com/unoplatform/uno.todo/blob/3895f7e9d1e3f5a15f229c0fb4229103e7c960b2/src/ToDo.UI.Tests/Given_MainPage.cs#L11

For context, the Welcome screen has two versions based on the size of the available screen, and each has a different "Get Started" button.

**Comment by agaldanaw2** [comment]
> WelcomePage_Button_Narrow
@Xiaoy312 
I run again with the change but I got same result:
image
Can you help me please?

File: src/ToDo/Presentation/WelcomeViewModel.cs (hunk 2 above)
Comment (agaldanaw2) [reply src/ToDo/Presentation/WelcomeViewModel.cs]: I upload the changes, can you check again, please?

**Comment by nickrandolph** [comment]
I've just separated the user and auth services so there's no cross dependencies. I think this branch needs to be rebased off main and then make sure all the unit and ui tests are passing

File: src/ToDo/Business/IUserProfilePictureService.cs [hunk 3]
```diff
@@ -3,5 +3,5 @@ namespace ToDo.Business;
 
 public interface IUserProfilePictureService
 {
-	Task<byte[]> GetAsync(CancellationToken ct);
+	ValueTask<byte[]> GetAsync(string userEmailAddress, CancellationToken cancellationToken);
```
Comment (carldebilly) [inline src/ToDo/Business/IUserProfilePictureService.cs:6]: It's unclear for me why the email would be required there...

**Review by carldebilly — APPROVED** [review]

File: src/ToDo/Business/IUserProfilePictureService.cs (hunk 3 above)
Comment (nickrandolph) [reply src/ToDo/Business/IUserProfilePictureService.cs:6]: We check to make sure that the profile picture is for this user. I guess we could have just passed in the usercontext object instead

**Comment by agneszitte-nventive** [comment]
@agaldanaw2 please if you can take care of the CLA also for this PR https://github.com/unoplatform/uno.todo/pull/259#issuecomment-1121461844

File: src/ToDo/Business/UserProfilePictureService.cs [hunk 4]
```diff
@@ -5,23 +5,20 @@ public class UserProfilePictureService : IUserProfilePictureService
...
 	private byte[]? _profilePicture;
-	private string? _userEmailAddress;
 
 	public UserProfilePictureService(IUserProfilePictureEndpoint client)
 	{
 		_client = client;
 	}
 
-	public async ValueTask<byte[]> GetAsync(string userEmailAddress, CancellationToken ct)
+	public async ValueTask<byte[]> GetAsync(CancellationToken ct)
```
Comment (nickrandolph) [inline src/ToDo/Business/UserProfilePictureService.cs:14]: @agaldanaw2 why was this change made? The emailAddress was there to make sure the cache is flushed if the signed in user changes?

**Review by nickrandolph — CHANGES_REQUESTED** [review]
Please revert the change that removes the emailAddress. This could be changed to UserContext but you need to pass something in to make sure the correct profile picture is being returned

File: src/ToDo/Business/UserProfilePictureService.cs (hunk 4 above)
Comment (nickrandolph) [reply src/ToDo/Business/UserProfilePictureService.cs:14]: cc @carldebilly @agneszitte-nventive

Comment (agneszitte-nventive) [reply src/ToDo/Business/UserProfilePictureService.cs]: Resolving the comment as you already made the proper changes to fix it

File: src/ToDo/Presentation/WelcomeViewModel.cs [hunk 5]
```diff
@@ -19,7 +19,6 @@ public partial class WelcomeViewModel
 	public async ValueTask GetStarted(CancellationToken ct)
 	{
 		var user = await _authService.AuthenticateAsync(_dispatcher);
```
Comment (agneszitte-nventive) [inline src/ToDo/Presentation/WelcomeViewModel.cs:21]: If there is no changes for WelcomeViewModel @nickrandolph maybe we can just remove it for this PR

**Review by agneszitte-nventive — DISMISSED** [review]

**Review by agneszitte-nventive — APPROVED** [review]

**Review by nickrandolph — APPROVED** [review]

**Review by agneszitte-nventive — APPROVED** [review]

**Pull request merged**